File: Singleton/Exercicios/Ex1_Singleton/Ex1.py
"""
PADRÃO: SINGLETON E PROTOTYPE (EXERCÍCIO SIGA)

Este exercício combina o Singleton para garantir uma única instância do sistema SIGA
e o Prototype para permitir a clonagem eficiente de registros de Alunos.
"""

import copy

# 1. PROTÓTIPO DE ALUNO
class Aluno:
    def __init__(self, curso, nome, periodo, idade, unidade="Fatec Diadema"):
        self.curso = curso
        self.nome = nome
        self.periodo = periodo
        self.idade = idade
        self.unidade = unidade

    # Método de clonagem (Prototype)
    def clone(self):
        return copy.copy(self)

    def __str__(self):
        return f"Aluno: {self.nome}, Curso: {self.curso}, Período: {self.periodo}, Idade: {self.idade}, Unidade: {self.unidade}"

# 2. SINGLETON SIGA (Gerenciador do Sistema)
class Siga:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(Siga, cls).__new__(cls)
            cls._instance.alunos = []
        return cls._instance

    @classmethod
    def get_instance(cls):
        return cls()

    def adicionar_aluno(self, aluno):
        self.alunos.append(aluno)

    def listar_alunos(self):
        return "\n".join(str(aluno) for aluno in self.alunos)

# 3. CLIENTE
if __name__ == "__main__":
    siga = Siga.get_instance()

    # Criando aluno original
    aluno1 = Aluno("DSM", "Stela", "Manhã", 19)
    siga.adicionar_aluno(aluno1)

    # Clonando aluno (Prototype)
    aluno2 = aluno1.clone()
    aluno2.nome = "João"
    siga.adicionar_aluno(aluno2)

    print("Alunos cadastrados no SIGA:")
    print(siga.listar_alunos())

File: Singleton/Exercicios/Ex1_Singleton/Ex1.js
/**
 * PADRÃO: SINGLETON E PROTOTYPE (EXERCÍCIO SIGA)
 * 
 * Este exercício demonstra o uso do Singleton para gerenciar o sistema SIGA
 * e o Prototype para clonar objetos de Alunos.
 */

// 1. PROTÓTIPO DE ALUNO
class Aluno {
    constructor(curso, nome, periodo, idade, unidade) {
        this.curso = curso;
        this.nome = nome;
        this.periodo = periodo;
        this.idade = idade;
        this.unidade = unidade || "Fatec Diadema";
    }

    // Método de clonagem (Prototype)
    clone() {
        return new Aluno(this.curso, this.nome, this.periodo, this.idade, this.unidade);
    }

    toString() {
        return `Aluno: ${this.nome}, Curso: ${this.curso}, Período: ${this.periodo}, Idade: ${this.idade}, Unidade: ${this.unidade}`;
    }
}

// 2. SINGLETON SIGA (Gerenciador do Sistema)
class Siga {
    constructor() {
        if (Siga.instance) {
            return Siga.instance;
        }
        this.alunos = [];
        Siga.instance = this;
    }

    static getInstance() {
        if (!Siga.instance) {
            new Siga();
        }
        return Siga.instance;
    }

    adicionarAluno(aluno) {
        this.alunos.push(aluno);
    }

    listarAlunos() {
        return this.alunos.map(a => a.toString()).join("\n");
    }
}

// 3. CLIENTE
const siga = Siga.getInstance();

// Criando aluno original
const aluno1 = new Aluno("DSM", "Stela", "Manhã", 19);
siga.adicionarAluno(aluno1);

// Clonando aluno para criar um novo (Prototype)
const aluno2 = aluno1.clone();
aluno2.nome = "João";
siga.adicionarAluno(aluno2);

console.log("Alunos cadastrados no SIGA:");
console.log(siga.listarAlunos());
